sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -327,16 +323,7 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]]
           val existingAttrMappingSet = transferAttrMapping.map(_._2).toSet
           newValidAttrMapping.filterNot { case (_, a) => existingAttrMappingSet.contains(a) }
         }
-        val resultAttrMapping = if (canGetOutput(plan)) {
-          // We propagate the attributes mapping to the parent plan node to update attributes, so
-          // the `newAttr` must be part of this plan's output.
```

Comment:
This logic itself is correct. An operator can only refer to an attribute if it's present in its child's output. I'm not sure how we produce such a plan
```
Filter (name#1 ...)
  Aggregate ([id#2], group by id#2)
    Scan [name#1, id#2]
```
The filter refers to `name` attribute but `Aggregate` does not output it. Can we investigate it further?

A safer fix can be: `DeduplicateRelations` should only dedup self-join if the two join children are resolved and have no missing attributes.

Reply:
As for why there is a problem with the view, there is no problem with the table because:
```
'Join Inner, ('l1.id = 'l2.id)
:- SubqueryAlias l1
: +- SubqueryAlias spark_catalog.default.t
: +- Relation default.t[id#20,name#21] parquet
+- SubqueryAlias l2
    +- !Filter (count(distinct tempresolvedcolumn(name#23, name)) > cast(1 as bigint))
       +- Aggregate [id#22], [id#22]
          +- SubqueryAlias spark_catalog.default.t
             +- Relation default.t[id#22,name#23] parquet
```
For tables, the id and name in join are different.
So it will not be considered duplicate and will not execute the following code.

Reply:
I was not asking why table is fine but view hits the bug. I was asking why we produce an invalid Filter operator. Even with the table case, you can also see `!Filter (...`, and the `!` indicates invalid plan. This seems like a feature of DataFrame API and IIUC the rule `ResolveMissingReferences` can fix the invalid `Filter` operator.

So my suggestion is: `DeduplicateRelations` should only kick in if the plan's children are all resolved and valid. e.g.
```
val newPlan = renewDuplicatedRelations(mutable.HashSet.empty, plan)._1
if (newPlan.find(p => p.resolved && p.missingInput.nonEmpty).isDefined) {
  // Wait for `ResolveMissingReferences` to resolve missing attributes first
  newPlan
} else {
  newPlan.resolveOperatorsUpWithPruning... {
    case p: LogicalPlan if !p.childrenResolved => p
      // To resolve duplicate expression IDs for Join.
      case j @ Join(left, right, _, _, _) if !j.duplicateResolved =>
        j.copy(right = dedupRight(left, right))
      // Resolve duplicate output for LateralJoin.
      ...
  }
}
```

Reply:
Thanks! I know how to do this.